evennia/utils/utils.py

```diff
@@ -2160,3 +2161,120 @@ def decorator(*args, **kwargs):
+    skip = False  # Skip the current parameter?
+    escape = False  # Was the escape key used?
+    is_string = False  # Have we been inside a quoted string?
+    temp = ""  # Buffer
+    key = None  # Key (for named parameter)
+
+    def _parse_value(temp):
+        ret = temp.strip()
+        if not is_string:
+            try:
+                ret = int(ret)
+            except ValueError:
+                try:
+                    ret = float(ret)
+                except ValueError:
+                    if ret != "":
+                        return FunctionArgument(ret)
+
+        return ret
+
+    def _add_value(skip, key, args_list, args_dict, temp):
+        if not skip:
+            # Record value based on whether named parameters mode is set or not.
+            if key is not None:
+                args_dict[key] = _parse_value(temp)
+                key = None
+            else:
+                args_list.append(_parse_value(temp))
+
+    for c in s:
+        if c == _ARG_ESCAPE_SIGN:
+            # Escape sign used.
+            if escape:
+                # Already escaping: print escape sign itself.
+                temp += _ARG_ESCAPE_SIGN
+                escape = False
+            else:
+                # Enter escape mode.
+                escape = True
+        elif escape:
+            # Escape mode: print whatever comes after the symbol.
+            escape = False
+            temp += c
+        elif inside[0] is True:
+            # Inside single quotes or double quotes
+            # Wait for the end symbol, allow everything else through, allow escape sign for typing quotes in strings
+            if c == inside[1]:
+                # Leaving single/double quoted area
+                inside = (False, None)
+            else:
+                temp += c
+        elif c == "\"" or c == "'":
+            # Entering single/double quoted area
+            inside = (True, c)
+            is_string = True
+            continue
+        elif c == "=":
+            if is_string:
+                # Invalid syntax because we don't allow named parameters to be quoted.
+                return None
+            elif key is None:
+                # Named parameters mode and equals sign encountered. Record key and continue with value.
+                key = temp.strip()
+                temp = ""
+        elif c == ",":
+            # Comma encountered outside of quoted area.
+
+            _add_value(skip, key, args_list, args_dict, temp)
+
+            # Reset
+            temp = ""
+            skip = False
+            is_string = False
+            key = None
+        else:
+            # Any other character: add to buffer.
+            temp += c
+
+    if inside[0] is True:
+        # Invalid syntax because we are inside a quoted area.
+        return None
+    else:
+        _add_value(skip, key, args_list, args_dict, temp)
+
+    return args_list, args_dict
```

Comment:
Some suggestions here:

1. > Don't compare boolean values to True or False using ==
[PEP 8 -- Programming Recommendations](https://www.python.org/dev/peps/pep-0008/#programming-recommendations)
2. >  When a string contains single or double quote characters, however, use the other one to avoid backslashes in the string. It improves readability.
[PEP-8 String Quotes](https://www.python.org/dev/peps/pep-0008/#string-quotes)
3. Return early, you will reduce the cognitive complexity of your function [cognitive complexity](https://www.sonarsource.com/docs/CognitiveComplexity.pdf)
```suggestion
    for c in s:
        if c == _ARG_ESCAPE_SIGN:
            # Escape sign used.
            if not escape:
                # Enter escape mode.
                escape = True
                continue
            # Already escaping: print escape sign itself.
            temp += _ARG_ESCAPE_SIGN
            escape = False
            continue

        if escape:
            # Escape mode: print whatever comes after the symbol.
            escape = False
            temp += c
            continue

        if inside[0]:
            # Inside single quotes or double quotes
            # Wait for the end symbol, allow everything else through, allow escape sign for typing quotes in strings
            if c != inside[1]:
                temp += c
                continue
            # Leaving single/double quoted area
            inside = (False, None)
            continue

        if c == '"' or c == "'":
            # Entering single/double quoted area
            inside = (True, c)
            is_string = True
            continue

        if c == "=":
            if is_string:
                # Invalid syntax because we don't allow named parameters to be quoted.
                return None
            if key is None:
                # Named parameters mode and equals sign encountered. Record key and continue with value.
                key = temp.strip()
                temp = ""
            continue

        if c == ",":
            # Comma encountered outside of quoted area.
            _add_value(skip, key, args_list, args_dict, temp)

            # Reset
            temp = ""
            skip = False
            is_string = False
            key = None
            continue
        # Any other character: add to buffer.
        temp += c

    if inside[0]:
        # Invalid syntax because we are inside a quoted area.
        return None
    _add_value(skip, key, args_list, args_dict, temp)
    return args_list, args_dict
```